File: homework01/vigener.go
package vigener

func EncryptVigenere(plaintext string, keyword string) string {
	var ciphertext string

	// PUT YOUR CODE HERE
	var sdvig int
	ciphertext = ""
	for i := 0; i < len(plaintext); i++ {
		symbol := int(plaintext[i])
		if (int('A') <= symbol) && (symbol <= int('Z')) || (int('a') <= symbol && symbol <= int('z')) {
			sdvig = int(keyword[i%len(keyword)])
			if (int('z') >= symbol) && (symbol >= int('a')) {
				sdvig -= int('a')
			} else {
				sdvig -= int('A')
			}

			symbol_code := symbol + sdvig
			if (int('a') <= symbol && symbol <= int('z')) && (symbol_code > int('z')) {
				symbol_code -= 26
			} else if int('A') <= symbol && symbol <= int('Z') && symbol_code > int('Z') {
				symbol_code -= 26
			}
			ciphertext += string(symbol_code)
		} else {
			ciphertext += string(symbol)
		}
	}

	return ciphertext
}

func DecryptVigenere(ciphertext string, keyword string) string {
	var plaintext string

	// PUT YOUR CODE HERE
	var sdvig int
	plaintext = ""
	for i := 0; i < len(ciphertext); i++ {
		symbol := int(ciphertext[i])
		if (int('A') <= symbol) && (symbol <= int('Z')) || (int('a') <= symbol && symbol <= int('z')) {
			sdvig = int(keyword[i%len(keyword)])
			if (int('z') >= symbol) && (symbol >= int('a')) {
				sdvig -= int('a')
			} else {
				sdvig -= int('A')
			}
			symbol_code := symbol - sdvig
			if (int('a') <= symbol && symbol <= int('z')) && (symbol_code < int('a')) {
				symbol_code += 26
			} else if int('A') <= symbol && symbol <= int('Z') && symbol_code < int('A') {
				symbol_code += 26
			}
			plaintext += string(symbol_code)
		} else {
			plaintext += string(symbol)
		}
	}

	return plaintext
}

File: homework01/vigener.py
def encrypt_vigenere(plaintext, keyword):
    """
    >>> encrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> encrypt_vigenere("python", "a")
    'python'
    >>> encrypt_vigenere("ATTACKATDAWN", "LEMON")
    'LXFOPVEFRNHR'
    """
    ciphertext = ""
    for num, symbol in enumerate(plaintext):
        if 'A' <= symbol <= 'Z' or 'a' <= symbol <= 'z':
            shift = ord(keyword[num % len(keyword)])
            shift -= ord('a') if 'z' >= symbol >= 'a' else ord('A')
            symbol_code = ord(symbol) + shift
            if 'a' <= symbol <= 'z' and symbol_code > ord('z'):
                symbol_code -= 26
            elif 'A' <= symbol <= 'Z' and symbol_code > ord('Z'):
                symbol_code -= 26
            ciphertext += chr(symbol_code)
        else:
            ciphertext += symbol
    return ciphertext


def decrypt_vigenere(ciphertext, keyword):
    """
    >>> decrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> decrypt_vigenere("python", "a")
    'python'
    >>> decrypt_vigenere("LXFOPVEFRNHR", "LEMON")
    'ATTACKATDAWN'
    """
    plaintext = ""
    for num, symbol in enumerate(ciphertext):
        if 'A' <= symbol <= 'Z' or 'a' <= symbol <= 'z':
            shift = ord(keyword[num % len(keyword)])
            shift -= ord('a') if 'z' >= symbol >= 'a' else ord('A')
            symbol_code = ord(symbol) - shift
            if 'a' <= symbol <= 'z' and symbol_code < ord('a'):
                symbol_code += 26
            elif 'A' <= symbol <= 'Z' and symbol_code < ord('A'):
                symbol_code += 26
            plaintext += chr(symbol_code)
        else:
            plaintext += symbol
    return plaintext
